hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala

```diff
@@ -0,0 +1,2259 @@
+  final def memoryManagedConsume(outerRegion: Value[Region], cb: EmitCodeBuilder, setup: EmitCodeBuilder => Unit = _ => ())(perElement: EmitCodeBuilder => Unit): Unit = {
+
+    if (separateRegions)
+      cb.assign(elementRegion, Region.stagedCreate(Region.REGULAR, outerRegion.getPool()))
+    else
+      cb.assign(elementRegion, outerRegion)
+
+    this.initialize(cb)
+    setup(cb)
+
+    cb.goto(this.LproduceElement)
+    cb.define(this.LproduceElementDone)
+    perElement(cb)
+
+    if (separateRegions)
+      cb += elementRegion.clearRegion()
+
+    cb.goto(this.LproduceElement)
+
+    cb.define(this.LendOfStream)
+    this.close(cb)
+
+    if (separateRegions)
+      cb += elementRegion.invalidate()
+  }
+}
+
+object EmitStream {
+  private[ir] def produce(
+    emitter: Emit[_],
+    streamIR: IR,
+    cb: EmitCodeBuilder,
+    outerRegion: Value[Region],
+    env: Emit.E,
+    container: Option[AggContainer]
+  ): IEmitCode = {
+
+    val mb = cb.emb
+
+
+    def emitVoid(ir: IR, cb: EmitCodeBuilder, region: Value[Region] = outerRegion, env: Emit.E = env, container: Option[AggContainer] = container): Unit =
+      emitter.emitVoid(cb, ir, region, env, container, None)
+
+    def emit(ir: IR, cb: EmitCodeBuilder, region: Value[Region] = outerRegion, env: Emit.E = env, container: Option[AggContainer] = container): IEmitCode = {
+      ir.typ match {
+        case _: TStream => produce(ir, cb, outerRegion, env, container)
+        case _ => emitter.emitI(ir, cb, region, env, container, None)
+      }
+    }
+
+    def produce(streamIR: IR, cb: EmitCodeBuilder, region: Value[Region] = outerRegion, env: Emit.E = env, container: Option[AggContainer] = container): IEmitCode =
+      EmitStream.produce(emitter, streamIR, cb, region, env, container)
+
+    streamIR match {
+
+      case NA(_typ) =>
+        val eltType = streamIR.pType.asInstanceOf[PCanonicalStream].elementType
+        val st = SStream(eltType.sType, false)
+        val region = mb.genFieldThisRef[Region]("na_region")
+        val producer = new StreamProducer {
+          override def initialize(cb: EmitCodeBuilder): Unit = {}
+
+          override val length: Option[Code[Int]] = Some(Code._fatal[Int]("tried to get NA stream length"))
+          override val elementRegion: Settable[Region] = region
+          override val separateRegions: Boolean = false
+          override val LproduceElement: CodeLabel = mb.defineHangingLabel { cb =>
+            cb.goto(LendOfStream)
+          }
+          override val element: EmitCode = EmitCode.missing(mb, eltType)
+
+          override def close(cb: EmitCodeBuilder): Unit = {}
+        }
+        IEmitCode.missing(cb, SStreamCode(st, producer))
+
+      case Ref(name, _typ) =>
+        assert(_typ.isInstanceOf[TStream])
+        env.lookup(name).toI(cb)
+          .map(cb) { case (stream: SStreamCode) =>
+            val childProducer = stream.producer
```

Comment:
Why can't you just return the child producer?

Reply:
This has to do with the unused label fixup. Consider:
```
StreamMap
  <some child stream<stream<T>>
  foo
  Ref foo
```

If we just return the child producer, then our label cleanup routine defines its labels, and the actual consumer (the node on top of the streamMap) tries to define the labels again.

Reply:
Doesn't the IR generating the stream define unused labels if there are no uses?  There is nobody else to do it, because if there are no references, the generator of the stream is the only place anything can happen.

 > If we just return the child producer, then our label cleanup routine defines its labels

I don't understand this.  Who's running the child clean routine in this example?  I won't hold up the PR for this discussion but I feel like there's something I'm missing.

Reply:
> Doesn't the IR generating the stream define unused labels if there are no uses?

Stream producers leave labels for their consumers to define. If we're emitting a StreamRange, for instance, there's no way for that node to know that its stream is unused and fix it up.

Instead, stream consumers must define all labels. We define a consumer as any IR with a [nested] stream child:
```
SomeIR
  <nested stream IR>
  ...
```

In this example, when we emit `SomeIR` (either as a realizable IEmitCode or as a stream producer IEmitCode), all labels, including nested, of `<nested stream IR>, must be defined. In the case of a `StreamMap` where we map a nested stream and the `StreamMap` body does not consume the inner stream (`(StreamMap <some_nested_stream> foo I32(1))` for instance), it is the responsibility of the `StreamMap` emitter to define the labels of the nested stream.

The subtlety comes in defining "does not consume the inner stream". I want a definition where any reference to the nested stream in `StreamMap` does consume the stream -- Including `(StreamMap some_nested "foo" (Ref "foo"))`. This makes `Ref` a consumer of a stream bound in the environment, which we implement by creating a new producer shimming the bound one.